extensions/s3/src/main/java/io/deephaven/extensions/s3/S3CompletableOutputStream.java

```diff
@@ -268,61 +313,102 @@ private void sendPartRequest(final OutgoingRequest request) throws IOException {
                 .partNumber(nextPartNumber)
                 .build();
         request.buffer.flip();
-        request.future = s3AsyncClient.uploadPart(uploadPartRequest,
-                AsyncRequestBody.fromByteBufferUnsafe(request.buffer));
         request.partNumber = nextPartNumber;
+        request.sdkUploadFuture = s3AsyncClient.uploadPart(uploadPartRequest,
+                AsyncRequestBody.fromByteBufferUnsafe(request.buffer))
+                .whenComplete((uploadPartResponse, throwable) -> {
+                    try {
+                        if (throwable == null) {
+                            completedParts.add(
+                                    SdkPojoConversionUtils.toCompletedPart(uploadPartResponse, request.partNumber));
+                        } else {
+                            failAll(new IOException("Failed to upload part " + request.partNumber + " for uri " + uri,
+                                    throwable));
+                        }
+                    } finally {
+                        request.releaseBuffer();
+                        numPartsCompleted.release();
+                    }
+                });
+
+        // If the status future fails, we want to cancel each part's upload
+        status.exceptionally(t -> {
+            request.sdkUploadFuture.cancel(true);
```

Comment:
Let's keep sdkUploadFuture as a local variable, so we don't need to implicitly keep hard reference to `request`.